### PR TITLE
Add S/Kademlia support to the stateless refactor.

### PR DESCRIPTION
**Pull request by doug-perlin**

### CODE REVIEW

File: base/messages.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+	if !bytes.Equal(recvRemote, remote) {
+		conn.Close()
+		return nil, errors.Errorf("inconsistent remotes %s and %s", hex.EncodeToString(recvRemote), hex.EncodeToString(remote))
+	}
+
+	if len(remoteAddr) > 255 {
```
Comment (jack0) [inline base/messages.go]: this and the length check on line 127 can be done after the local/remote length check

File: base/messages.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+func runRecvWorker(finalizerNotifier chan struct{}, conn net.Conn, callback protocol.RecvMessageCallback) {
+	reader := bufio.NewReader(conn)
+
+	for {
+		// conn should also be closed as soon as gcFinalize() is called
+		// so we do not need to check finalizerNotifier?
+		n, err := binary.ReadUvarint(reader)
+		if err != nil {
+			break
```
Comment (jack0) [inline base/messages.go]: should we log the error here and on line 204?

File: protocol/node.go [hunk 3]
```diff
@@ -2,52 +2,91 @@ package protocol
...
+
+func (n *Node) GetIdentityAdapter() IdentityAdapter {
+	return n.idAdapter
+}
+
+func (n *Node) GetConnectionAdapter() ConnectionAdapter {
+	return n.connAdapter
+}
+
+func (n *Node) ManuallyRemovePeer(remote []byte) {
```
Comment (jack0) [inline protocol/node.go]: why don't we just make removePeer RemovePeer?

File: .gitignore [hunk 4]
```diff
@@ -9,6 +9,4 @@ coverage.txt
 
 noise
 chat
-debug
-
-go.sum
```
Comment (doug-perlin) [inline .gitignore]: Not a fan of this go.sum, it doesn't add any benefit and will be different for every developer. Should keep it out.

Comment (iwasaki-kenta) [reply .gitignore]: https://github.com/golang/go/wiki/Modules#should-i-commit-my-gosum-file-as-well-as-my-gomod-file

Comment (doug-perlin) [reply .gitignore]: That link has been referenced in several PR's, but the claims seem incredibly weak to me. They basically don't trust github at google, so other projects have to suffer the consequences.

File: protocol/node.go (hunk 3 above)
Comment (doug-perlin) [reply protocol/node.go]: I'll remove it

**Pull request merged**